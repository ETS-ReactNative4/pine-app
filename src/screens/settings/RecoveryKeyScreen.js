import React, { Component } from 'react';
import { Alert, Linking } from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ReactNativeHaptic from 'react-native-haptic';
import iCloudAccountStatus from 'react-native-icloud-account-status';

import * as keyActions from '../../actions/keys';
import getMnemonicByKey from '../../crypto/getMnemonicByKey';
import headerStyles from '../../styles/headerStyles';
import BackButton from '../../components/BackButton';
import SettingsAttribute from '../../components/SettingsAttribute';
import SettingsButton from '../../components/SettingsButton';
import SettingsGroup from '../../components/SettingsGroup';
import SettingsDescription from '../../components/SettingsDescription';
import BaseSettingsScreen from './BaseSettingsScreen';

const OPEN_SETTINGS_URL_STRING = 'App-Prefs:';

@connect((state) => ({
  keys: state.keys.items,
  hasCreatedBackup: state.settings.user.hasCreatedBackup,
  pineAddress: state.settings.user.profile.address
}))
export default class RecoveryKeyScreen extends Component {
  static navigationOptions = ({ navigation }) => ({
    title: 'Recovery Key',
    headerStyle: headerStyles.header,
    headerTitleStyle: headerStyles.title,
    headerLeft: (<BackButton onPress={() => { navigation.goBack(); }} />)
  });

  state = {
    phrase: '',
    storeInICloud: false
  }

  componentDidMount() {
    const dispatch = this.props.dispatch;
    const keys = Object.values(this.props.keys);
    const defaultKey = keys[0];

    getMnemonicByKey(defaultKey.id).then((mnemonic) => {
      this.setState({ phrase: mnemonic });
      this._checkICloudBackupStatus();
    });
  }

  _checkICloudBackupStatus() {
    const dispatch = this.props.dispatch;
    const mnemonic = this.state.phrase;

    return iCloudAccountStatus.getStatus().then((accountStatus) => {
      if (accountStatus !== iCloudAccountStatus.STATUS_AVAILABLE) {
        return;
      }

      return dispatch(keyActions.recover()).then((recoveredMnemonic) => {
        if (recoveredMnemonic !== mnemonic) {
          return;
        }

        this.setState({ storeInICloud: true });
      });
    });
  }

  _updateICloudState(storeInICloud) {
    const { dispatch, pineAddress } = this.props;
    const mnemonic = this.state.phrase;

    this.setState({ storeInICloud });

    if (storeInICloud) {
      return dispatch(keyActions.backup(mnemonic, pineAddress));
    }

    return dispatch(keyActions.removeBackup(pineAddress));
  }

  _onStoreInICloudChange(storeInICloud) {
    const activate = !this.state.storeInICloud && storeInICloud;

    if (!activate && !this.props.hasCreatedBackup) {
      return this._showMissingManualBackupAlert();
    }

    if (activate) {
      return iCloudAccountStatus.getStatus().then((accountStatus) => {
        if (accountStatus === iCloudAccountStatus.STATUS_AVAILABLE) {
          return this._updateICloudState(storeInICloud);
        }

        this._showICloudUnavailableAlert();
      });
    }

    return this._updateICloudState(storeInICloud);
  }

  _showMissingManualBackupAlert() {
    ReactNativeHaptic.generate('notificationWarning');

    Alert.alert(
      'Missing Manual Backup',
      'The recovery key can only be removed from iCloud if it has been manually backed up.',
      [
        { text: 'Cancel', style: 'cancel' },
        {
          text: 'Back up Manually',
          onPress: this._backUpManually.bind(this)
        }
      ],
      { cancelable: false }
    );
  }

  _showICloudUnavailableAlert() {
    ReactNativeHaptic.generate('notificationWarning');

    Alert.alert(
      'iCloud Not Available',
      'Make sure that you are logged in to iCloud and has given Pine access.',
      [
        { text: 'Cancel', style: 'cancel' },
        {
          text: 'Settings',
          onPress: () => { Linking.openURL(OPEN_SETTINGS_URL_STRING); }
        }
      ],
      { cancelable: false }
    );
  }

  _backUpManually() {
    const navigation = this.props.navigation;

    navigation.navigate('BackUpMnemonic', {
      mnemonic: this.state.phrase,
      isModal: true
    });
  }

  render() {
    const hasCreatedBackup = this.props.hasCreatedBackup;

    return (
      <BaseSettingsScreen>
        <SettingsGroup>
          <SettingsAttribute
            name='Back up in iCloud'
            value={this.state.storeInICloud}
            onValueChange={this._onStoreInICloudChange.bind(this)}
            isLastItem={true}
          />
        </SettingsGroup>
        <SettingsDescription>
          Saving your recovery key in iCloud is safe and convenient as long as you keep your iCloud account secure.
        </SettingsDescription>

        <SettingsGroup>
          <SettingsButton
            title='Back up Manually'
            showCheckmark={Boolean(hasCreatedBackup)}
            onPress={this._backUpManually.bind(this)}
            isLastItem={true}
          />
        </SettingsGroup>
        <SettingsDescription>
          Backing up your recovery key manually on a piece of paper is a good alternative in case you would
          lose access to your iCloud account, or if you don&#39;t want to store your key in iCloud.
        </SettingsDescription>
      </BaseSettingsScreen>
    );
  }
}

RecoveryKeyScreen.propTypes = {
  dispatch: PropTypes.func,
  navigation: PropTypes.any,
  keys: PropTypes.object,
  hasCreatedBackup: PropTypes.bool,
  pineAddress: PropTypes.string
};
