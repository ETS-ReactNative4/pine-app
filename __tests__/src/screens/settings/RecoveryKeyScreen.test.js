import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import RecoveryKeyScreen from '../../../../src/screens/settings/RecoveryKeyScreen';

jest.mock('../../../../src/containers/ErrorModalContainer', () => 'ErrorModalContainer');
jest.mock('../../../../src/containers/MnemonicWordsContainer', () => 'MnemonicWordsContainer');

const storeMock = {
  getState: jest.fn(() => ({
    settings: {
      user: {
        hasCreatedBackup: false,
        profile: {
          address: 'test@pine.dev'
        }
      }
    },
    keys: {
      items: {
        'b62a06e9-7a1f-4862-a350-7af4505114c7': {
          id: 'b62a06e9-7a1f-4862-a350-7af4505114c7'
        }
      }
    }
  })),
  dispatch: jest.fn(),
  subscribe: jest.fn()
};

const navigationMock = {
  navigate: jest.fn(),
  state: {
    params: {}
  }
};

describe('RecoveryKeyScreen', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <RecoveryKeyScreen store={storeMock} navigation={navigationMock} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
