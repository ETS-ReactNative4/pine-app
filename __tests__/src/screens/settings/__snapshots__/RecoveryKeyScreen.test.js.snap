// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecoveryKeyScreen renders correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "backgroundColor": "#EFEFF3",
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "paddingTop": 35,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "borderBottomColor": "#C8C8CC",
                "borderBottomWidth": 0.5,
                "borderTopColor": "#C8C8CC",
                "borderTopWidth": 0.5,
                "marginBottom": 35,
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderBottomColor": "#C8C7CC",
                  "borderBottomWidth": 0.5,
                  "flex": 1,
                  "height": 44,
                  "justifyContent": "center",
                  "marginLeft": 15,
                  "paddingRight": 15,
                },
                Object {
                  "borderBottomWidth": 0,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "fontSize": 17,
                    "fontWeight": "400",
                    "lineHeight": 22,
                  },
                ]
              }
            >
              Back up in iCloud
            </Text>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#8F8E94",
                    "fontSize": 17,
                    "fontWeight": "400",
                    "lineHeight": 22,
                    "position": "absolute",
                    "right": 15,
                    "textAlign": "right",
                    "width": 250,
                  },
                  Object {
                    "width": null,
                  },
                ]
              }
            >
              <RCTSwitch
                onChange={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "height": 31,
                    "width": 51,
                  }
                }
                value={false}
              />
            </View>
          </View>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "color": "#8A8A8F",
                "fontSize": 13,
                "lineHeight": 17,
                "marginBottom": 30,
                "marginTop": -25,
                "paddingLeft": 15,
                "paddingRight": 15,
              },
            ]
          }
        >
          Saving your recovery key in iCloud is safe and convenient as long as you keep your iCloud account secure.
        </Text>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "borderBottomColor": "#C8C8CC",
                "borderBottomWidth": 0.5,
                "borderTopColor": "#C8C8CC",
                "borderTopWidth": 0.5,
                "marginBottom": 35,
              },
              undefined,
            ]
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "borderBottomColor": "#C8C7CC",
                    "borderBottomWidth": 0.5,
                    "flex": 1,
                    "height": 44,
                    "justifyContent": "center",
                    "marginLeft": 15,
                    "paddingRight": 15,
                  },
                  Object {
                    "borderBottomWidth": 0,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Array [
                      Object {
                        "fontSize": 17,
                        "fontWeight": "400",
                        "lineHeight": 22,
                      },
                      Object {
                        "color": "#007AFF",
                      },
                      undefined,
                      undefined,
                    ],
                  ]
                }
              >
                Back up Manually
              </Text>
            </View>
          </View>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "color": "#8A8A8F",
                "fontSize": 13,
                "lineHeight": 17,
                "marginBottom": 30,
                "marginTop": -25,
                "paddingLeft": 15,
                "paddingRight": 15,
              },
            ]
          }
        >
          Backing up your recovery key manually on a piece of paper is a good alternative in case you would lose access to your iCloud account, or if you don't want to store your key in iCloud.
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <ErrorModalContainer />
</View>
`;
